### DISCUSSION TITLE
where did using lifetime steps go???

### DISCUSSION THREAD

**Discussion by alenknight** [d8ahazard/sd_dreambooth_extension#673] [General]
i was using "save every x amount of lifetime steps" to generate iterative ckpt.... now it's per epoch... only.  i can't seem to find the option anymore in the latest update where to NOT use epochs... anyone have an idea?
(epochs confuse me... sorry, it's just so much easier to use lifetime steps)

Also I set it to save after every epoch. 20 source images and at 10,000 steps now it's still not saving iteratively.

**Comment by d8ahazard** [comment]
They're gone, and it's not coming back.

Lifetime steps is, with all honesty, a very confusing and poor measure of actual training progress. In the original scripts, a "step" was just one "fetch" of training data from the dataloader. How many images are fetched? It depends on the batch size. If we want to get even more confusing, gradient accumulation theoretically acts as a multiplier to the total number of "iterations" that happen over a given image. Maybe.

Epochs, while still somewhat ephemeral, are at least a value that has some correspondence to the overall length of the dataset being trained on, as they *always* equivalate to "one iteraton over the whole dataset".

I legitimately do not know or understand why "steps" became the default measure of how much a model was trained, but if you think about it, it provides no actual information on the extent of the training. Reading the original HuggingFace blog on Dreambooth training is a great example of this. "We trained our model for 100 steps at a LR of xxx". Yeah, but how many instance images? With steps, there's always either uncertainty or math required to understand how much training was actually done.

Epochs? There is no math. There is no guessing. If you say you did 100 epochs, I instantly know you did 100 steps per image. What size is your dataset? Doesn't especially matter. 

As far as saving - take a look at the new tab of the same name, especially the "Generate xxx *during* training." settings. This affects what files get generated when. By default, the general diffusion weights are *always* saved at the specified interval during training. I'd expect you to be able to look in the /working directory of your model folder and see that the unet/tenc .bin files have more recent update dates than their corresponding config files. 

Only when you check the checkboxes for "save a checkpoint during training" or "save a checkpoint when training completes" will training create the specific .ckpt files.

**Reply by andykaufseo** [reply]
totally understand why you did this, it makes more sense this way, and it's the way it should be

**Reply by alenknight** [reply]
thanks for explaining... and that makes sense now why using Epochs... guess i need to start getting comfy with calculating them :)

Only issue now - it seems to not be saving as you mentioned.  so i go to Saving tab, and i select  "save a checkpoint during training" or "save a checkpoint when training completes".... 
and it works when i first start up the first training session.  but if i cancel anytime, or start a new one.... the settings no longer work.
i'm testing with 4 images/40 Classification images... and set training steps per image to 100.
save preview and ckpt both at 3 (for testing).
first run, works fine.
if i hit cancel....or even if i finish and create a new training model... it doesn't seem to stick.  
has this been an issue for others?

**Comment by 1025307336** [comment]
Save Model Frequency (Epochs)，Save Preview(s) Frequency (Epochs)，It seems that it has become a multiple multiplication based on a certain number. I generate a preview every 1800 steps, and generate a ckpt every 9000 steps.

**Reply by andykaufseo** [reply]
it's based on epochs, which are based on how many training images you have.
if you have 100 images and do 10 epochs, that's 1000 steps.
if you have 1000 images and do 1 epoch, that's also 1000 steps.

training steps are irrelevant, the important metric is epochs, meaning how many times each image is trained.

if you have 70 images and want to do 1000 steps, some images will get trained more than others - that's why steps are not a good way to measure / set things - and epochs are the right way to do it.

**Comment by FugueSegue** [comment]
For the last day I have been trying to figure out why Dreambooth is acting in an unexpected way. I haven't had this trouble before. I thought something was broken on my installation. It's been happening since the newest Dreambooth extension update. So I reinstalled Ubuntu, Auto1111, and the Dreambooth extension and started from scratch. Same result.

70 dataset images
100 epochs
Batch Size 1
Gradient Accumulation Steps 1

The training goes for 14,000 steps. I see it in the CLI. The resulting checkpoint generates a technicolor yawn.
The training should be only 7,000 steps. I can't find anything in the GUI that would double this number. I am absolutely baffled. What am I missing?
